src/libos/crates/async-rt/src/sync/mutex/inner.rs

```diff
@@ -0,0 +1,221 @@
+///                Thread 1              |                     Thread 2
+///  no lock:        Free                |  no lock:         (do something)
+///  try lock:      Locked               |  no lock:         (do something)
+///  hold lock:   (do something)         |  try lock:       LockedWithWaiters
+///  hold lock:   (do something)         |  wait lock:           (sleep)
+///  unlock:         Free                |  wait lock:           (sleep)
+///  no lock:    (do something else)     |  wake and try lock:   Locked
+///  no lock:    (do something else)     |  hold lock:        (do something)
+///  no lock:    (do something else)     |  unlock:               Free
+#[derive(Debug)]
+pub(super) struct MutexInner {
+    status: AtomicLockStatus,
+    waiters: AtomicI32,
+    // Although WaiterQueue has inner mutex, we still need an exterior mutex to synchronize the WaiterQueue and the status.
+    waiter_queue: Mutex<WaiterQueue>,
+}
+
+// This struct is the atomic wrapper for LockStatus.
+#[derive(Debug)]
+struct AtomicLockStatus(AtomicI32);
+
+// The status code keeps the same as the implementation in musl.
+#[derive(Debug, Copy, Clone)]
+#[repr(i32)]
+enum LockStatus {
+    Free = 0,
+    Locked = Errno::EBUSY as i32,
+    LockedWithWaiters = -2147483632, // 0x8000_0010 = (EBUSY | 0x8000_0000)
+}
+
+impl MutexInner {
+    pub(super) fn new() -> MutexInner {
+        MutexInner {
+            status: AtomicLockStatus::init(),
+            waiters: AtomicI32::new(0),
+            waiter_queue: Mutex::new(WaiterQueue::new()),
+        }
+    }
+
+    pub(super) fn try_lock(&self) -> Result<()> {
+        if self.status.try_set_lock().is_ok() {
+            Ok(())
+        } else {
+            Err(errno!(EBUSY, "the lock is held by other threads"))
+        }
+    }
+
+    pub(super) async fn lock(&self) -> Result<()> {
+        if let Ok(_) = self.try_lock() {
+            return Ok(());
+        }
+
+        // Spin for a short while if no one is waiting but the lock is held.
+        let mut spins = SPIN_COUNT;
+        while spins != 0
+            && self.status.is_locked()
+            // Can't reorder here. `Relaxed` is enough.
+            && self.waiters.load(Ordering::Relaxed) == 0
+        {
+            hint::spin_loop();
+            spins -= 1;
+        }
+
+        loop {
+            if let Ok(_) = self.try_lock() {
+                return Ok(());
+            }
+
+            if self.status.is_free() {
+                continue;
+            }
+
+            // In try_set_lock_with_waiters, `AcqRel` will make sure this increment happens before. Thus, `Relaxed` can be used here.
+            self.waiters.fetch_add(1, Ordering::Relaxed);
+
+            // Ignore the result here. If the state transition fails, the next wait will not block.
+            self.status
+                .try_set_lock_with_waiters()
+                .map_err(|e| errno!(e.errno(), "failed to set lock status"))
+                .ok();
```

Comment:
If results are to be ignored, then the right to do it is as follows.

```rust
let _ = self.status.try_set_lock_with_waiters();
```

In addition, printing errors and warnings is not a good idea. Because failures to grab a lock are harmless, trivial, and may be very frequent in some cases. There is no point in printing such messages. They are going to be noises.

Reply:
Done.